### PR TITLE
refactor: migrate from windi to tailwind

### PR DESCRIPTION
**Pull request by elevatebart**
*Edit by @lmiller1990* WindiCSS (Tailwind clone) is official deprecated, but has been unmaintained for a while now. @elevatebart updated us to use the official Tailwind. We need to review the Launchpad and App to make sure it all still looks correct.

Steps:

- [x] Upgrade Design System items: `npx npm-check-updates "/^@cypress-design/*/" -wsu`
- [x] follow the [instructions to switch from cypress-windi to tailwind](https://design.cypress.io/Install.html#tailwindcss)
- [x] fix all custom values incompatible with tailwind
  - [x] fix margin and padding classes:`(-?)([pm][xyltbr]?)-([\d.]+)(px|em|rem|vh)` -> `$2-[$1$3$4]` from `p-16px` to `p-[16px]`
  - [x] position classes: `(-?)(left|right|top|bottom)-([\d.]+)(px|em|rem|vh)` -> `$2-[$1$3$4]`
  - [x] text size classes: `(text|leading)-([\d.]+)(px|em|rem|vh)` -> `$1-[$2$3]`
    - [x] `text-space-([\d.]+)(px|em|rem|vh)` -> `text-[$1$2]`
  - [x] border: `(border|outline|divider|ring)(-[xyltrb])?-([\d.]+)(px|em|rem|vh)` -> `$1$2-[$3$4]`
  - [x] dimensions: `(min-|max-)?([wh])-([\d.]+)(px|em|rem|vh)` -> `$1$2-[$3$4]`
  - [x] `(gap|space)(-[xy])?-(\d+)px` -> `$1$2-[$3px]`
  - [x] `rounded(-[tb][lr])?-(\d+)px` -> `rounded$1-[$2px]` // rounded
  - [x] `(-?)translate(-[xy])?-(\d+)px` -> `translate$2-[$1$3px]` // translate
- [x] Some classes are different
  - [x] `flex-grow`, `flex-shrink` and `flex-gap` should be `grow`, `shrink` and `gap`
  - [x] `underline-indigo-500` should be `decoration-indigo-500`
  - [x] `border-1` should be `border`
    - [x] `border(-[x|y|t|b|l|r|s|e])-1` -> `border$1`	
- [x] remove all grouped variants `hover:(bg-indigo-500 underline text-white)`  does not work with tailwind
- [x] add back the `children:` variant
- [x] fix some defaults about pseudo class `before:` (add `before:shrink-0`)

### Fixed redundant styles for component tests
* For `packages/app` and `packages/launchpad`, the utility classes were getting loaded twice.  Once was when the component support file was importing the `frontend-shared` component support file, and then again added directly in the project's component support file.  This PR just relies on the imports in the `frontend-shared` component support file. 

### Monospace font
* Moved the import of the 'Fire Code' monospace font back to the `shared.scss` file and using the `woff2` version versus the `ttf` version
  * Removed the inclusion of the `ttf` font in the `App.vue` files for `package/app` and `package/launchpad`

### Key differences when using Tailwind vs Windi
* Our Windi implementation applied `!important` to all the utility classes.  Tailwind currently does not do this.  This would cause a different precedence when having overlapping styles.  
  * Can cause a difference hen two utility classes for the same style were added to a DOM element.
  * In Tailwind, you can add a `!` prefix to a utility class to get Tailwind to dynamically add a `!important`
* In Windi, if you used a dynamic text size (i.e. `text-16px` vs `text-sm`), then Windi also included `line-height: 1` as part of that style.  TailWind does not do this. Instead, Tailwind inherits the line-height for these elements. The majority of differences in the Percy snapshots for this PR that will be introduced are related to this.  The snapshots will show slight increases in line height in this case.
* Updating the way that the monospaced font is included

### CODE REVIEW

**Comment by lmiller1990** [comment]
We will review this next sprint @elevatebart

**Comment by lmiller1990** [comment]
Bumping this into our current sprint for review.

File: packages/app/package.json [hunk 1]
```diff
@@ -20,8 +20,9 @@
   },
   "dependencies": {},
   "devDependencies": {
-    "@cypress-design/vue-icon": "0.18.1",
-    "@cypress-design/vue-statusicon": "0.2.4",
+    "@cypress-design/css": "0.11.0",
```
Comment (warrensplayer) [inline packages/app/package.json]: Why does `@cypress-design/css` need to be here as a dependency?  It is used via frontend-shared

File: packages/launchpad/package.json [hunk 2]
```diff
@@ -21,6 +20,7 @@
   },
   "dependencies": {},
   "devDependencies": {
+    "@cypress-design/css": "0.11.0",
```
Comment (warrensplayer) [inline packages/launchpad/package.json]: Does this need to be a dependency here?

File: packages/app/cypress/component/support/ctSupport.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import { AutIframe } from '../../../src/runner/aut-iframe'
 import { EventManager } from '../../../src/runner/event-manager'
 import type { Socket } from '@packages/socket/lib/browser'
+import '../../../src/main.scss'
```
Comment (warrensplayer) [inline packages/app/cypress/component/support/ctSupport.ts]: Why is this line here when it was not needed before?

**Review by warrensplayer — COMMENTED** [review]
Reviewed the PR description and it all makes sense.  I checked out the branch and started spot checking a few things.  I  committed a few updates:
* changed `text-size-*` styles to `text-*`
* had to add `!` to the beginning of a few styles in the `InlineSpecListHeader` to override styles from the normalized styles. I know the Windi styles were run through some post processor to add `!important` to all of them, so not sure what else might be a fall out from not having that
* Removed duplicate text color style from `SpecItem`. Tailwind has those styles in a different order, so a different one was taking precedence.  My IDE was complaining that there were duplicate styles on the tag I fixed.  Is there a way to "lint" the tailwind styles to find issues like this?

The biggest issue right now is the font family.  It is different now with Tailwind than it was for Windi. That appears to be the main cause of all the current Percy snapshot differences.

File: packages/launchpad/package.json (hunk 2 above)
Comment (elevatebart) [reply packages/launchpad/package.json]: nah, I think you are right It was a peace of mind step for me.
I will clean up.

File: packages/app/package.json (hunk 1 above)
Comment (elevatebart) [reply packages/app/package.json]: nah, I think you are right It was a peace of mind step for me.
I will clean up.

File: packages/app/cypress/component/support/ctSupport.ts (hunk 3 above)
Comment (elevatebart) [reply packages/app/cypress/component/support/ctSupport.ts]: We needed to import tailwind and I figured it was a mistake that the rest of the shared css was not imported here.

I will remove this change and replace it by 

```js
import 'tailwindcss/tailwind.css`
```

Comment (elevatebart) [reply packages/app/cypress/component/support/ctSupport.ts]:
```suggestion
import 'tailwindcss/tailwind.css'
```

Comment (elevatebart) [reply packages/app/cypress/component/support/ctSupport.ts]: this will probably accelerate the tests

Comment (elevatebart) [reply packages/app/cypress/component/support/ctSupport.ts]: actually, it was a mistake...

**Comment by warrensplayer** [comment]
@elevatebart Just an FYI on some changes I just pushed:
* I reverted the change in `DebugRunNavigation` you made yesterday back to a `before` pseudo element to remove the need for a separate `li`
* Fixed a few more cases of the Windi  grouped styles
* added a custom `rotate` option for 270deg that was used in a few places

Are you looking at the font-family issue I noted in my initial review?  I can take a look later on today but wanted to check with you first so not to duplicate work.  I think once this is solved, then the Percy snapshots will be much more useful for any remaining fixes we need to make.

**Comment by warrensplayer** [comment]
@elevatebart fixes from this morning
* fixed some border styles for values matching the regex `border-([x|y|t|b|l|r|s|e])-1` and updated them to `border-$1`
* couple other small fixes
* update some of the documentation

Today, can you look at the following: 
* updating the other documentation that references "Windi"
* Can we get rid of the Windi dependencies in the `@cypress-design/css` module?  It is still pulling those in.

**Comment by cbfrance** [comment]
@elevatebart will this PR change any of the default breakpoints or are they the same between windi and tailwind?

**Comment by elevatebart** [comment]
> @elevatebart will this PR change any of the default breakpoints or are they the same between windi and tailwind?

WindiCSS uses precisely the same default breakpoint as tailwind. No change necessary

**Comment by elevatebart** [comment]
Wow! So much good work!
There is enough for a blogpost here! 

Thank you @warrensplayer for taking this on.

**Comment by lmiller1990** [comment]
I found some minor issues I could easily fix. There are a bunch of "off by a few pixels" which don't look too concerning. I think getting this 1:1 pixel perfect is going to be too much work. I'll see which quick wins I can get today, and once this is in a good enough place and we are confidence, we can merge it up.

**Comment by warrensplayer** [comment]
In the last 8 commits, please review the following:

* I removed redundant style imports for CT tests.  They were already being imported when the component support file from frontend-shared was being imported.
  * packages/app/cypress/component/support/index.ts
  * packages/launchpad/cypress/component/support/index.ts
* I put the WOFF2 version of the Fira Code font back in the scss file and removed the TTF version from the App.vue files in packages/app and packages/launchpad.  I am not a font expert, but as I understand it, the WOFF2 font is the more modern font that contains multiple variants.  The options in the @font-face definition are ones that I just pulled back in that were there previously.  This is the cause for a lot of the Percy differences. The app should now always use this font for the monospace font option.  Since it has multiple variants in it, there are now some places where a different font weight was being applied and it will now be honored.
  * packages/app/src/App.vue
  * packages/launchpad/src/App.vue
  * packages/frontend-shared/src/styles/shared.scss

The other changes are 
* removing unneeded `!` overrides since the styles are now ordered correctly
* Fixing some places that were affected by Windi previously always applying `line-height: 1` when a specific test size was specified in pixels
* Inlining some sample code into the `packages/frontend-shared/src/components/ShikiHighlight.cy.tsx` test

**Comment by lmiller1990** [comment]
This is looking good to me. Some of the component test specific snapshots have slightly different spacing, but all the ones in E2E looks either very similar, or when the padding is a little different, I think they still look fine.

**Review by lmiller1990 — APPROVED** [review]

**Review by warrensplayer — APPROVED** [review]

**Comment by marktnoonan** [comment]
I'm taking a walk through Percy to approve it, curious about a couple screems:

1. Settings container differences (new snapshots have different overflow - doesn't seem to be a problem, just different)

https://percy.io/cypress-io/cypress/builds/26992166/changed/1501872371?browser=chrome&browser_ids=38&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=original&width=900&widths=300%2C500%2C580%2C600%2C700%2C800%2C900%2C1000%2C1032%2C1119%2C1200%2C1280

https://percy.io/cypress-io/cypress/builds/26992166/changed/1501872390?browser=chrome&browser_ids=38&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=900&widths=300%2C500%2C580%2C600%2C700%2C800%2C900%2C1000%2C1032%2C1119%2C1200%2C1280

2. This icon is upside down now

Screenshot 2023-04-28 at 8 30 18 AM

 https://percy.io/cypress-io/cypress/builds/26992166/changed/1501872332?browser=chrome&browser_ids=38&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=800&widths=300%2C500%2C580%2C600%2C700%2C800%2C900%2C1000%2C1032%2C1119%2C1200%2C1280

**Comment by warrensplayer** [comment]
@marktnoonan The overflow is due to the app now consistently using the same monospace font.  The WOFF version of the font seems to be slightly wider.

I looked into that upside down icon. I am not sure why it was not upside down before when looking at the code.  The code seems to expect that an expandable alert will always use a caret icon that should rotate when opening/closing.

**Comment by marktnoonan** [comment]
Got it @warrensplayer, I've gone and approved Percy 👍
